src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -208,4 +213,61 @@ public String deleteControlledGcsBucket(
 
     return jobBuilder.submit();
   }
+
+  /**
+   * Users who are READERS on the workspace may only have READER  on a private resource.
+   * WRITER workspace users may have either READER or WRITER assigned to them.
+   */
+  private void validateUserWorkspaceRole() {
+
+  }
+
+  private void validateOnlySelfAssignmentForWorkspaceWriter(ControlledResource controlledResource, AuthenticatedUserRequest userRequest) {
+    if (!controlledResource.getAccessScope().equals(AccessScopeType.ACCESS_SCOPE_PRIVATE)) {
+      // No need to handle SHARED resources
+      return;
+    }
+    final boolean userIsWriterOnWorkspace = samService.isAuthorized(
```

Comment:
owners also have "write" permission on a workspace. It's probably easier to check `userIsWorkspaceOwner` and flip some conditions here.